### ISSUE TITLE
Controller extension

### ISSUE THREAD

**Issue by rodrigodalcindev** [sprangular/sprangular#175]
Ideally, we should be able to extend controllers instead of overriding them in our Sprangular-based apps, in order to allow for customization and decrease eventual conflicts with framework upgrades.

**Comment by joshnuss** [comment]
There are 2 scenarios:

1) Sprangular's controller does exactly what I need, except for one minor thing
    The host app should override the route, and create a new controller which extends Sprangular's:

```coffeescript
module.controller 'CustomProductCtrl', ($scope, $controller) ->
    # Initialize the super class and extend it.
    angular.extend(this, $controller('ProductCtrl', {$scope: $scope}))
    
   # Additional extensions to create a mixin.
   this.icanhaz = ->
      alert "icanhaz"

```

2) Sprangular's controller doesnt do what I need:
    The host app should define its own controller, and override the route

**Comment by DanielWright** [comment]
What are real-world use-cases for extending a controller?

**Comment by rodrigodalcindev** [comment]
@DanielWright I think the product controller on ADAY is a good example. We could test out @joshnuss solution on a separate branch.

**Comment by DanielWright** [comment]
I don't agree. I think that controller is an example of shoving stuff to a controller that should have been refactored elsewhere. There's DOM manipulation that should be in directives, there's data-model stuff that should be in models.

Controllers, IMO, should be for setting up initial scope and nothing else.

**Comment by simonwalsh** [comment]
@DanielWright Agreed. But this means that we should then have an issue to convert controllers to directives. It is one or the other IMO.

And then, would there be a neat way to extend directives?

**Comment by joshnuss** [comment]
Extending things partially should always be a last resort. Ideally things should be factored well so that parts are designed to be replaced.

In other words, if a controller is getting too complex, it should be split up into parts, so that apps can use the parts they want and toss the stuff they dont. 

Can we do a review of all controllers overwritten across our projects and figure out what the reasons were for overriding in each case?

@DanielWright controllers should be about setting up initial scope *and* provide functions to communicate with services. Its the middleman. The HTML shouldn't really be dependent on anything other than the controller.
